File: gimbal/src/gimbal_node.cpp
// ROS 头文件
#include <ros/ros.h>
// 用于发送云台控制包的头文件
#include <mavros_msgs/MountControl.h>
// 用于接受云台的姿态包的头文件
#include <geometry_msgs/Quaternion.h>
// 四元数和RPY欧拉角转换函数头文件
#include <tf/tf.h>

// 定义一些必要的全局变量
double pitch, roll, yaw;
bool gimbal_init = false;
// ROS 订阅来自地面站的目标选择消息的回调函数
void mount_orientation_cb(const geometry_msgs::Quaternion::ConstPtr& msg){
    gimbal_init = true;
    // 四元数到RPY欧拉角转换,注意极性
    double pitch_r, roll_r, yaw_r;
    tf::Quaternion quater;
    tf::quaternionMsgToTF(*msg, quater);
    tf::Matrix3x3(quater).getRPY(roll_r, pitch_r, yaw_r);
    pitch = pitch_r * 180 / M_PI;
    roll = roll_r * 180 / M_PI;
    yaw = -yaw_r * 180 / M_PI ;
    ROS_INFO("Received Mount Orientaion Message pitch, roll, yaw: (%f, %f, %f)", pitch, roll, yaw);
}

int main(int argc, char **argv)
{
    // 初始化 ROS 节点
    ros::init(argc, argv, "gimbal_node");
    ros::NodeHandle nh;

    // ROS 订阅获取云台的姿态信息
    ros::Subscriber mount_orientation_sub = nh.subscribe<geometry_msgs::Quaternion>
            ("mavros/mount_control/orientation", 1, mount_orientation_cb);
    // ROS Topic 发布控制云台姿态
    ros::Publisher mount_control_pub = nh.advertise<mavros_msgs::MountControl>
            ("mavros/mount_control/command", 1);

    // 设定值发布速率 20 hz
    ros::Rate rate(5.0);

    // 等待 Gimbal 姿态包
    while(ros::ok() && !gimbal_init){
        ros::spinOnce();
        rate.sleep();
    }

    // 定义飞机解锁消息包
    mavros_msgs::MountControl mount_control;
    // 保持 Roll 为 0
    mount_control.roll = 0;
    // 角度模式
    mount_control.mode = 2;
    // 控制角度变化
    mount_control.yaw = fabs(yaw) < 10 ? 100 : 0;
    mount_control.pitch = 0;
    // 主循环
    while(ros::ok()) {
        mount_control_pub.publish(mount_control);

        ros::spinOnce();
        rate.sleep();
    }

    return 0;
}

File: gimbal/scripts/gimbal_node.py
#! /usr/bin/env python

import math
import rospy
from geometry_msgs.msg import Quaternion
from mavros_msgs.msg import MountControl

def quaternion_to_euler(x, y, z, w):
    t0 = +2.0 * (w * x + y * z)
    t1 = +1.0 - 2.0 * (x * x + y * y)
    X = math.degrees(math.atan2(t0, t1))

    t2 = +2.0 * (w * y - z * x)
    t2 = +1.0 if t2 > +1.0 else t2
    t2 = -1.0 if t2 < -1.0 else t2
    Y = math.degrees(math.asin(t2))

    t3 = +2.0 * (w * z + x * y)
    t4 = +1.0 - 2.0 * (y * y + z * z)
    Z = math.degrees(math.atan2(t3, t4))

    return X, Y, Z

pitch = 0.0
roll = 0.0
yaw = 0.0
gimbal_init = False

def mount_orientation_cb(msg):
    global gimbal_init,pitch,roll,yaw
    gimbal_init = True
    roll,pitch,yaw = quaternion_to_euler(msg.x, msg.y, msg.z, msg.w)
    yaw=-yaw
    rospy.loginfo("Received Mount Orientaion Message pitch, roll, yaw: (%f, %f, %f)", pitch, roll, yaw)

if __name__ == "__main__":
    rospy.init_node("gimbal_node_py")

    mount_orientation_sub = rospy.Subscriber("mavros/mount_control/orientation", Quaternion, callback = mount_orientation_cb)
    mount_control_pub = rospy.Publisher("mavros/mount_control/command", MountControl, queue_size=20)

    rate = rospy.Rate(5)
    
    while(not rospy.is_shutdown() and not gimbal_init):
        rate.sleep()

    mount_control = MountControl()
    mount_control.roll = 0
    mount_control.mode = 2
    if math.fabs(yaw) < 10:
        mount_control.yaw = 100
    else:
        mount_control.yaw = 0
    mount_control.pitch = 0;

    while(not rospy.is_shutdown()):
        mount_control_pub.publish(mount_control)

        rate.sleep()